README.md
# Admin Expert Mode

A plugin for WordPress that allows users to hide inline documentation and help text that are geared for beginning users in the WordPress admin.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/admin-expert-mode/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Each user who wishes to enable expert mode for themselves must do so individually by going into their profile, checking the checkbox 'Expert mode', and then pressing the 'Update Profile' button.


## Additional Documentation

* See [readme.txt](https://github.com/coffee2code/admin-expert-mode/blob/master/readme.txt) for additional usage information.
* See [DEVELOPER-DOCS.md](DEVELOPER-DOCS.md) for developer-related documentation on hooks, template tags, and shortcode.
* See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/admin-expert-mode/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/admin-expert-mode/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/admin-expert-mode/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 2.9 _(2021-11-16)_

### Highlights:

This minor release adds DEVELOPER-DOCS.md, notes compatibility through WP 5.8+, reorganizes unit tests, and minor tweaks.

### Details:

* New: Add DEVELOPER-DOCS.md and move hooks documentation into it
* Change: Add newlines after output block tags and remove newline after label tag
* Change: Add translator comments to explain placeholders
* Change: Use stricter equality check in a conditional statement
* Change: Tweak inline function documentation (typo, verb tenses, bullet list syntax)
* Change: Tweak installation instruction
* Change: Note compatibility through WP 5.8+
* Change: Change a tag in readme.txt header
* Unit tests:
    * Change: Use stricter regex in a couple tests
    * Change: Restructure unit test file structure
        * Change: Move `tests/test-*` into `tests/phpunit/tests/`
        * Change: Move `tests/bootstrap.php` to `tests/phpunit/`
        * Change: Move `bin/` into `phpunit/`
    * Change: In bootstrap, store path to plugin file constant so its value can be used within that file and in test file
    * Change: In bootstrap, add backcompat for PHPUnit pre-v6.0
    * Change: Remove 'test-' prefix from unit test files
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* New: Add a few more possible TODO items

## 2.8 _(2021-03-24)_

### Highlights:

This release is a recommended minor update that hides text in the welcome panel on the dashboard page, hides text on the "Settings - Writing" and "Settings - Reading" pages, and notes compatibility through WP 5.7+.

### Details:

* New: Hide the description of the dashboard's welcome panel
* New: Hide the label for the "Update Services" on the "Settings - Writing" page
* New: Hide the extra description for the "Search engine visibility" on the "Settings - Reading" page
* Fix: Fix typo in plugin description
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 2.7 _(2020-06-14)_

### Highlights:

This release is a recommended minor update that updates hiding of text on "Privacy Settings" page, hides text on "Media Settings" page, hides text relating to avatars on "Discussion Setting" page, adds a TODO.md file, updates a few URLs to be HTTPS, expands unit testing, and notes compatibility through WP 5.4+.

### Details:

* New: Hide the descriptions of what avatars and default avatars are on the "Discussion Settings" page
* New: Hide the intro paragraph for the "Media Settings" page
* New: Add TODO.md and move existing TODO list from top of main plugin file into it
* Fix: Re-hide descriptive paragraphs for "Privacy Settings" page
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Unit tests:
    * Change: Enhance `test_default_hooks()` to support testing for direct function callbacks rather than just method callbacks
    * Change: Remove unnecessary unregistering of hooks
    * Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and delete commented-out code)

## 2.6 _(2019-12-22)_

### Highlights:

This release is a minor update that fixes a bug with considering the user setting value, verifies compatibility through WordPress 5.3+, adds unit testing, and makes minor behind-the-scenes improvements.

### Details:

* Fix: Properly account for user setting value
* New: Add unit testing
* New: Add `reset()` to reset plugin to pre-pageload state
* Change: Clarify the inline docs for `c2c_admin_expert_mode_default` filter to indicate it won't override user setting value
* Change: Remove memoization of `$is_active` value
* Change: Allow class to be defined even when loaded outside the admin
* Change: Note compatibility through WP 5.3+
* Change: Add inline documentation for class variables
* Change: Add link to CHANGELOG.md in README.md
* Change: Update copyright date (2020)
* Delete: Remove unused class variable `$config`

## 2.5 _(2019-04-04)_

### Highlights:

This minor release primarily adds the ability for admin users to see and edit the setting within other users' profiles and notes compatilibity through WordPress v5.1+. All other changes were behind-the-scenes for the general improvement of the plugin and its documentation.

### Details:

* New: Permit admins to see and edit the value of the setting for other users
* Change: Add user ID as an additional argument to the `c2c_admin_expert_mode_default` filter
* Change: Allow `get_options()` to accept a user ID and memoize values by user ID
* Change: Initialize plugin on `plugins_loaded` action instead of on load
* Change: Merge `do_init()` into `init()`
* Change: Modify help text for checkbox
* Change: Cast return value of `c2c_admin_expert_mode` filter as boolean
* Change: (Hardening) Escape output of user profile URL
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hooks
* Change: Improve function docblocks by adding missing `@see`, `@access`, `@param`, and `@return` tags
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS
* Change: Split paragraph in README.md's "Support" section into two

## 2.4 _(2018-04-13)_
* New: Add ability to disable expert mode for current display of a given page by appending '?disable-admin-expert-mode=1' to the URL
* New: Hide descriptive paragraphs for "Privacy Settings" page
* New: Hide description for "Biographical Info" field in user profile
* New: Hide description for "Sessions" field in user profile
* New: Add README.md
* Change: Delete the user option if the checkbox wasn't checked
* Change: Use `sprintf()` to format output markup rather than concatenating strings and variables
* Change: (Hardening) Check that current user is able to edit the user being edited before saving the user option

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Admin_Expert_Mode
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/admin-expert-mode.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Admin_Expert_Mode
 */

define( 'ADMIN_EXPERT_MODE_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/admin-expert-mode.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require ADMIN_EXPERT_MODE_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-admin-expert-mode.php
<?php

defined( 'ABSPATH' ) or die();

class Admin_Expert_Mode_Test extends WP_UnitTestCase {

	protected static $admin_options_name = 'c2c_admin_expert_mode';
	protected static $disable_query_key = 'disable-admin-expert-mode';
	protected static $field_name = 'admin_expert_mode';
	protected static $style_handle = 'c2c_AdminExpertMode_admin';
	protected static $transient = 'aem_activated';
	protected $user_id = 0;

	public function setUp() {
		parent::setUp();

		$this->user_id = $this->factory->user->create( array( 'role' => 'editor' ) );
		wp_set_current_user( $this->user_id );
	}

	public function tearDown() {
		parent::tearDown();

		c2c_AdminExpertMode::reset();

		delete_transient( self::$transient );
		delete_user_option( $this->user_id, self::$admin_options_name );
		$this->user_id = 0;
		unset( $_GET[ self::$disable_query_key ] );
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	public static function get_default_hooks() {
		return array(
			array( 'action', 'admin_init',               'register_styles',           10 ),
			array( 'action', 'admin_notices',            'display_activation_notice', 10 ),
			array( 'action', 'personal_options',         'show_option',               10 ),
			array( 'action', 'personal_options_update',  'maybe_save_options',        10 ),
			array( 'action', 'edit_user_profile_update', 'maybe_save_options',        10 ),
			array( 'action', 'admin_enqueue_scripts',    'enqueue_admin_css',         10 ),
		);
	}


	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_AdminExpertMode' ) );
	}

	public function test_get_version() {
		$this->assertEquals( '2.7', c2c_AdminExpertMode::version() );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_AdminExpertMode', 'init' ) ) );
	}

	/**
	 * @dataProvider get_default_hooks
	 */
	public function test_default_hooks( $hook_type, $hook, $function, $priority, $class_method = true ) {
		$callback = $class_method ? array( 'c2c_AdminExpertMode', $function ) : $function;

		$prio = $hook_type === 'action' ?
			has_action( $hook, $callback ) :
			has_filter( $hook, $callback );

		$this->assertNotFalse( $prio );
		if ( $priority ) {
			$this->assertEquals( $priority, $prio );
		}
	}

	public function test_plugin_activated() {
		$this->assertFalse( get_transient( self::$transient ) );

		c2c_AdminExpertMode::plugin_activated();

		$this->assertEquals( 'show', get_transient( self::$transient ) );
	}

	/*
	 * reset()
	 */

	public function test_reset() {
		// This assigns a value to a user which then gets cached.
		update_user_option( $this->user_id, self::$admin_options_name, true );
		$options = c2c_AdminExpertMode::get_options();

		$this->assertTrue( $options[ self::$field_name ] );

		// Change value.
		update_user_option( $this->user_id, self::$admin_options_name, false );
		c2c_AdminExpertMode::reset();
		$options = c2c_AdminExpertMode::get_options();

		$this->assertNotTrue( $options[ self::$field_name ] );
	}

	/*
	 * is_admin_expert_mode_active()
	 */

	public function test_default_value_for_is_admin_expert_mode_active() {
		$this->assertFalse( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_user_option_enables_is_admin_expert_mode_active() {
		update_user_option( $this->user_id, self::$admin_options_name, true );

		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_query_key_disables_is_admin_expert_mode() {
		$this->test_user_option_enables_is_admin_expert_mode_active();

		$_GET[ self::$disable_query_key ] = '1';

		$this->assertFalse( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	/*
	 * filter: c2c_admin_expert_mode
	 */

	public function test_filter_cc2c_admin_expert_mode_enables_mode() {
		add_filter( 'c2c_admin_expert_mode', '__return_true' );

		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_filter_cc2c_admin_expert_mode_disables_mode() {
		add_filter( 'c2c_admin_expert_mode', '__return_false' );

		update_user_option( $this->user_id, self::$admin_options_name, true );

		$this->assertFalse( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	/*
	 * filter: c2c_admin_expert_mode_default
	 */

	public function test_filter_c2c_admin_expert_mode_default_true_enables_mode() {
		add_filter( 'c2c_admin_expert_mode_default', '__return_true' );

		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_filter_c2c_admin_expert_mode_default_false_leaves_mode_disabled() {
		add_filter( 'c2c_admin_expert_mode_default', '__return_false' );

		$this->assertFalse( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_filter_c2c_admin_expert_mode_default_true_does_not_override_user_setting() {
		add_filter( 'c2c_admin_expert_mode_default', '__return_true' );

		update_user_option( $this->user_id, self::$admin_options_name, true );

		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	public function test_filter_c2c_admin_expert_mode_default_false_does_not_override_user_setting() {
		add_filter( 'c2c_admin_expert_mode_default', '__return_false' );

		update_user_option( $this->user_id, self::$admin_options_name, true );

		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
	}

	/*
	 * display_activation_notice()
	 */

	public function test_display_activation_notice_nothing_display_when_plugin_not_just_activated() {
		$this->expectOutputRegex( '/^$/', c2c_AdminExpertMode::display_activation_notice() );
	}

	public function test_display_activation_notice_displays_when_plugin_just_activated() {
		$this->test_plugin_activated();

		$this->expectOutputRegex( "/^<div id='message' class='updated fade'><p>.+<\/p><\/div>$/", c2c_AdminExpertMode::display_activation_notice() );
	}

	public function test_display_activation_notice_when_inactive() {
		$this->test_plugin_activated();

		$expected = '<strong>NOTE:</strong> You must enable expert mode for yourself (in your <a href="http://example.org/wp-admin/profile.php" title="Profile">profile</a>) for it to take effect. Other admin users must do the same for themselves as well. (See the readme.txt for more advanced controls.)';

		$this->expectOutputRegex(
			'|' . preg_quote( $expected ) . '|', c2c_AdminExpertMode::display_activation_notice() );
	}

	public function test_display_activation_notice_when_active() {
		$this->test_plugin_activated();
		update_user_option( $this->user_id, self::$admin_options_name, true );

		$expected = 'Expert mode is now enabled for you since you had it previously enabled. You can disable it in your <a href="http://example.org/wp-admin/profile.php" title="Profile">profile</a>. Reminder: other admins must separately enable expert mode for themselves via their own profiles. (See the readme.txt for more advanced controls.)';

		$this->expectOutputRegex( '|' . preg_quote( $expected ) . '|', c2c_AdminExpertMode::display_activation_notice() );
	}

	/*
	 * register_styles()
	 */

	public function test_register_styles() {
		$this->assertFalse( wp_style_is( self::$style_handle, 'registered' ) );

		c2c_AdminExpertMode::register_styles();

		$this->assertTrue( wp_style_is( self::$style_handle, 'registered' ) );

		// Cleanup.
		wp_deregister_style( self::$style_handle );
		$this->assertFalse( wp_style_is( self::$style_handle, 'registered' ) );
	}

	/*
	 * enqueue_admin_css()
	 */

	public function test_enqueue_admin_css() {
		c2c_AdminExpertMode::register_styles();
		c2c_AdminExpertMode::enqueue_admin_css();

		$this->assertTrue( wp_style_is( self::$style_handle, 'registered' ) );
		$this->assertFalse( wp_style_is( self::$style_handle, 'enqueued' ) );

		c2c_AdminExpertMode::reset();
		update_user_option( $this->user_id, self::$admin_options_name, true );
		$this->assertTrue( c2c_AdminExpertMode::is_admin_expert_mode_active() );
		c2c_AdminExpertMode::enqueue_admin_css();

		$this->assertTrue( wp_style_is( self::$style_handle, 'enqueued' ) );

		// Cleanup.
		wp_deregister_style( self::$style_handle );
		$this->assertFalse( wp_style_is( self::$style_handle, 'registered' ) );
		wp_dequeue_style( self::$style_handle );
		$this->assertFalse( wp_style_is( self::$style_handle, 'enqueued' ) );
	}

	/*
	 * show_option()
	 */

	public function test_show_option_with_setting_disabled() {
		$expected = '<tr><th scope="row"></th><td><label for="admin_expert_mode"><input type="checkbox" id="admin_expert_mode" name="admin_expert_mode" value="1">' . "\n" . '</label></td></tr>';

		$this->expectOutputRegex(
			'|' . preg_quote( $expected ) . '|',
			c2c_AdminExpertMode::show_option( wp_get_current_user() )
		);
	}

	public function test_show_option_with_setting_enabled() {
		update_user_option( $this->user_id, self::$admin_options_name, true );

		$expected = '<tr><th scope="row"></th><td><label for="admin_expert_mode"><input type="checkbox" id="admin_expert_mode" name="admin_expert_mode" value="1" checked=\'checked\'>' . "\n" . '</label></td></tr>';

		$this->expectOutputRegex(
			'|' . preg_quote( $expected ) . '|',
			c2c_AdminExpertMode::show_option( wp_get_current_user() )
		);
	}

	/*
	 * get_options()
	 */

	public function test_get_options_with_invalid_user() {
		$this->assertEquals( array( 'admin_expert_mode' => false ), c2c_AdminExpertMode::get_options( 99999 ) );
	}

	public function test_get_options_default_implied_current_user() {
		update_user_option( $this->user_id, self::$admin_options_name, true );

		$this->assertEquals( array( 'admin_expert_mode' => true ), c2c_AdminExpertMode::get_options() );
	}

	/*
	 * maybe_save_options()
	 */

	public function test_maybe_save_options_called_by_user_without_privilege() {
		$_POST[ self::$field_name ] = '1';

		$user_id = $this->factory->user->create( array( 'role' => 'subscriber' ) );
		wp_set_current_user( $user_id );

		c2c_AdminExpertMode::maybe_save_options( $user_id );

		$this->assertFalse( get_user_option( $user_id, self::$admin_options_name ) );
	}

	public function test_maybe_save_options_saves_setting_privileged_self() {
		$_POST[ self::$field_name ] = '1';

		c2c_AdminExpertMode::maybe_save_options( $this->user_id );

		$this->assertEquals( array( self::$field_name => true ), get_user_option( self::$admin_options_name, $this->user_id ) );
	}

	public function test_maybe_save_options_saves_setting_being_enabled() {
		$_POST[ self::$field_name ] = '1';

		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
		wp_set_current_user( $user_id );

		c2c_AdminExpertMode::maybe_save_options( $this->user_id );

		$this->assertEquals( array( self::$field_name => true ), get_user_option( self::$admin_options_name, $this->user_id ) );
	}

}


admin-expert-mode.php
<?php
/**
 * Plugin Name: Admin Expert Mode
 * Version:     2.9
 * Plugin URI:  https://coffee2code.com/wp-plugins/admin-expert-mode/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: admin-expert-mode
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Allows users to hide inline documentation and help text that are geared for beginning users in the WordPress admin.
 *
 * Compatible with WordPress 2.8 through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/admin-expert-mode/
 *
 * @package Admin_Expert_Mode
 * @author  Scott Reilly
 * @version 2.9
 */

/*
	Copyright (c) 2009-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_AdminExpertMode' ) ) :

class c2c_AdminExpertMode {

	/**
	 * The name for the admin option.
	 *
	 * @access private
	 * @var string
	 */
	private static $admin_options_name = 'c2c_admin_expert_mode';

	/**
	 * Name for the admin field.
	 *
	 * @access private
	 * @var string
	 */
	private static $field_name         = 'admin_expert_mode';

	/**
	 * Translated prompt/label for the admin field.
	 *
	 * Note: This is set during `init()`.
	 *
	 * @access private
	 * @var string
	 */
	private static $prompt             = '';

	/**
	 * Translated help text for the admin field.
	 *
	 * Note: This is set during `init()`.
	 *
	 * @access private
	 * @var string
	 */
	private static $help_text          = '';

	/**
	 * Memoized plugin configuration.
	 *
	 * @access private
	 * @var array
	 */
	private static $options            = array();

	/**
	 * Name of the query key used for disabling expert mode for current page.
	 *
	 * @since 2.4
	 * @var string
	 */
	private static $disable_query_key  = 'disable-admin-expert-mode';


	/**
	 * Returns version of the plugin.
	 *
	 * @access public
	 * @since 1.8
	 */
	public static function version() {
		return '2.9';
	}

	/**
	 * Initializes plugin.
	 *
	 * @access public
	 * @since 1.0.0
	 */
	public static function init() {
		// Fire off a function when the plugin gets activated.
		register_activation_hook( __FILE__, array( __CLASS__, 'plugin_activated' ) );

		if ( is_admin() ) {
			// Load textdomain.
			load_plugin_textdomain( 'admin-expert-mode' );

			// Set translatable strings.
			self::$prompt =    __( 'Expert mode', 'admin-expert-mode' );
			self::$help_text = __( 'Enable expert mode (disables most of the onscreen documentation in the admin)', 'admin-expert-mode' );
		}

		// Register and enqueue styles for admin page.
		add_action( 'admin_init',               array( __CLASS__, 'register_styles'           ) );

		// Display admin notice on same page load as plugin activation.
		add_action( 'admin_notices',            array( __CLASS__, 'display_activation_notice' ) );

		// Add the checkbox to user profiles.
		add_action( 'personal_options',         array( __CLASS__, 'show_option'               ) );

		// Save the user preference.
		add_action( 'personal_options_update',  array( __CLASS__, 'maybe_save_options'        ) );
		add_action( 'edit_user_profile_update', array( __CLASS__, 'maybe_save_options'        ) );

		// Enqueue admin scripts and styles.
		add_action( 'admin_enqueue_scripts',    array( __CLASS__, 'enqueue_admin_css'         ) );
	}

	/**
	 * Resets plugin to pre-pageload state.
	 *
	 * Note: This mostly clears runtime caches. It does not revert the plugin to
	 * its freshly installed state.
	 *
	 * @access public
	 * @since 2.6
	 */
	public static function reset() {
		self::$options = array();
	}

	/**
	 * Sets a temporary flag (transient) to indicate the plugin was just activated.
	 *
	 * @access public
	 * @since 1.5.0
	 */
	public static function plugin_activated() {
		set_transient( 'aem_activated', 'show', 10 );
	}

	/**
	 * Outputs activation notice.
	 *
	 * @access public
	 * @since 1.0.0
	 */
	public static function display_activation_notice() {
		if ( get_transient( 'aem_activated' ) ) {
			if ( self::is_admin_expert_mode_active() ) {
				/* translators: %s: URL to user's profile page in the admin. */
				$msg = __( 'Expert mode is now enabled for you since you had it previously enabled. You can disable it in your <a href="%s" title="Profile">profile</a>. Reminder: other admins must separately enable expert mode for themselves via their own profiles. (See the readme.txt for more advanced controls.)', 'admin-expert-mode' );
			} else {
				/* translators: %s: URL to user's profile page in the admin. */
				$msg = __( '<strong>NOTE:</strong> You must enable expert mode for yourself (in your <a href="%s" title="Profile">profile</a>) for it to take effect. Other admin users must do the same for themselves as well. (See the readme.txt for more advanced controls.)', 'admin-expert-mode' );
			}
			$msg = sprintf( $msg, esc_url( admin_url( 'profile.php' ) ) );

			echo "<div id='message' class='updated fade'><p>$msg</p></div>\n";
		}
	}

	/**
	 * Indicates if admin expert mode is active for the current user.
	 *
	 * Takes the following into account in this order:
	 * - Value of 'c2c_admin_expert_mode' filter, if true
	 * - Value of the per-user setting
	 *
	 * @access public
	 * @since 1.5.0
	 *
	 * @return bool True is admin expert mode is active, else false.
	 */
	public static function is_admin_expert_mode_active() {
		$options = self::get_options();

		if ( ! empty( $_GET[ self::$disable_query_key ] ) ) {
			return false;
		}

		/**
		 * Filters whether the admin expert mode should be active for the current
		 * user.
		 *
		 * @since 1.5.0
		 *
		 * @param bool   $is_active  Is admin expert mode currently enabled for user?
		 * @param string $user_login Login of the current user.
		 */
		return (bool) apply_filters( 'c2c_admin_expert_mode', $options[ self::$field_name ], get_user_option( 'user_login' ) );
	}

	/**
	 * Registers styles.
	 *
	 * @access public
	 * @since 1.8
	 */
	public static function register_styles() {
		wp_register_style( __CLASS__ . '_admin', plugins_url( 'admin.css', __FILE__ ) );
	}

	/**
	 * Enqueues stylesheets if the user has admin expert mode activated.
	 *
	 * @access public
	 * @since 1.8
	 */
	public static function enqueue_admin_css() {
		if ( self::is_admin_expert_mode_active() ) {
			wp_enqueue_style( __CLASS__ . '_admin' );
		}
	}

	/**
	 * Outputs the form input field for the admin expert mode setting checkbox.
	 *
	 * @access public
	 * @since 1.0.0
	 *
	 * @param WP_User $user The user whose options are being shown.
	 */
	public static function show_option( $user ) {
		$current_user = wp_get_current_user();
		$is_current_user_profile_page = ( $user->ID === $current_user->ID );

		// Only show on current user's own profile, or other user profiles if current
		// user has appropriate capabilities.
		if ( ! $is_current_user_profile_page && ! current_user_can( 'edit_users' ) ) {
			return;
		}

		$options = self::get_options( $user->ID );

		echo '<tr><th scope="row">' . self::$prompt . '</th>';
		echo '<td>';
		printf(
			'<label for="%s"><input type="checkbox" id="%s" name="%s" value="%s"%s>',
			esc_attr( self::$field_name ),
			esc_attr( self::$field_name ),
			esc_attr( self::$field_name ),
			'1',
			checked( (bool) $options[ self::$field_name ], true, false )
		);
		echo self::$help_text;
		echo "</label></td></tr>\n";
	}

	/**
	 * Returns array of the plugin's settings.
	 *
	 * @access public
	 * @since 1.0.0
	 *
	 * @param int $user_id The user ID. Default is ID of current user.
	 * @return array The plugin's settings
	 */
	public static function get_options( $user_id = 0 ) {
		if ( ! $user_id ) {
			$user_id = wp_get_current_user()->ID;
		}

		if ( ! empty( self::$options[ $user_id ] ) ) {
			return self::$options[ $user_id ];
		}

		$user_setting_value = get_user_option( self::$admin_options_name, $user_id );

		/**
		 * Filters whether the admin expert mode should be active by default.
		 *
		 * Note: This only affects the default value and does not override what the
		 * user has set themselves. To override the value of the user setting (and
		 * thus have the final say in whether the mode is enabled or not, use the
		 * `c2c_admin_expert_mode` filter.)
		 *
		 * @since 1.5.0
		 *
		 * @param bool $is_active Is admin expert mode active by default? Default false.
		 * @param int  $user_id   The user ID.
		 */
		$default = (bool) apply_filters( 'c2c_admin_expert_mode_default', false, $user_id );
		self::$options[ $user_id ][ self::$field_name ] = $user_setting_value || $default;

		return self::$options[ $user_id ];
	}

	/**
	 * Saves value of checkbox to allow user to opt into receiving
	 * notifications for all comments.
	 *
	 * @access public
	 * @since 1.0.0
	 *
	 * @param int  $user_id The user ID.
	 */
	public static function maybe_save_options( $user_id ) {
		if ( ! current_user_can( 'edit_user', $user_id ) ) {
			return false;
		}

		$options = self::get_options( $user_id );
		$options[ self::$field_name ] = ! empty( $_POST[ self::$field_name ] );

		if ( empty( $_POST[ self::$field_name ] ) ) {
			delete_user_option( $user_id, self::$admin_options_name );
		} else {
			update_user_option( $user_id, self::$admin_options_name, $options );
		}

		self::$options[ $user_id ] = $options;
	}

} // end c2c_AdminExpertMode

add_action( 'plugins_loaded', array( 'c2c_AdminExpertMode', 'init' ) );

endif; // end if !class_exists()
